tests/subsys/net/lib/nrf_cloud/src/main.c

```diff
@@ -0,0 +1,191 @@
+ * have a success initialization in the test below
+ */
+void event_handler(const struct nrf_cloud_evt *const evt)
+{
+	switch (evt->type) {
+	case NRF_CLOUD_EVT_READY:
+		printk("Device is connected and ready");
+		break;
+	case NRF_CLOUD_EVT_ERROR:
+		printk("Failed to connect to cloud");
+		break;
+	default:
+		break;
+	}
+}
+
+/* This function runs before each test */
+void setUp(void)
+{
+	/* Uninit to change the internal state to idle */
+	__wrap_nct_uninit_Ignore();
+	nrf_cloud_uninit();
+}
+
+/* This function runs after each completed test */
+void tearDown(void)
+{
+	/* Uninit to reset the internal state to idle */
+	__wrap_nct_uninit_Expect();
+	nrf_cloud_uninit();
+}
+
+/* The library can only be initialized when the internal state is
+ * idle, so the test verifies that nrf_cloud_init function will
+ * return with permission denied error (-EACCES) when not idle
+ */
+void test_init_not_idle(void)
+{
+	struct nrf_cloud_init_param params = {
+		.event_handler = event_handler,
+		.client_id = NULL,
+	};
+
+	TEST_ASSERT_EQUAL_MESSAGE(STATE_IDLE, nfsm_get_current_state(),
+		"internal state should be idle");
+
+	/* Call nrf_cloud_init twice with the first one
+	 * being successful and check if the second one
+	 * will fail since the state will not be idle
+	 */
+	__wrap_nct_init_IgnoreAndReturn(0);
+	nrf_cloud_init(&params);
+
+	int ret = nrf_cloud_init(&params);
+
+	TEST_ASSERT_NOT_EQUAL_MESSAGE(STATE_IDLE, nfsm_get_current_state(),
+		"internal state should not be idle");
+
+	TEST_ASSERT_EQUAL_MESSAGE(-EACCES, ret, "return should be -EACCES when not idle");
+}
+
+/* Verify that nrf_cloud_init function will throw an error
+ * of invalid argument (-EINVAL) when given null params
+ */
+void test_init_null_params(void)
+{
+	TEST_ASSERT_EQUAL_MESSAGE(STATE_IDLE, nfsm_get_current_state(),
+		"internal state should be idle");
+
+	int ret = nrf_cloud_init(NULL);
+
+	TEST_ASSERT_EQUAL_MESSAGE(-EINVAL, ret, "return should be -EINVAL with NULL params");
+}
+
+/* Verify that nrf_cloud_init function will throw an error
+ * of invalid argument (-EINVAL) when a null event handler
+ * is passed in as part of its parameters
+ */
+void test_init_null_handler(void)
+{
+	struct nrf_cloud_init_param params = {
+		.event_handler = NULL,
+		.client_id = NULL,
+	};
+
+	TEST_ASSERT_EQUAL_MESSAGE(STATE_IDLE, nfsm_get_current_state(),
+		"internal state should be idle");
+
+	int ret = nrf_cloud_init(&params);
+
+	TEST_ASSERT_EQUAL_MESSAGE(-EINVAL, ret,
+		"return should be -EINVAL with NULL event handler");
+}
+
+/* Verify that nrf_cloud_init function will succeed
+ * given valid internal state and event handler
+ */
+void test_init_success(void)
+{
+	struct nrf_cloud_init_param params = {
+		.event_handler = event_handler,
+		.client_id = NULL,
+	};
+
+	/* Expect nct_init to run successfully */
+	__wrap_nct_init_ExpectAndReturn(params.client_id, 0);
+
+	TEST_ASSERT_EQUAL_MESSAGE(STATE_IDLE, nfsm_get_current_state(),
```

Comment:
I suggest putting this assertion before the call to `__wrap_nct_init_ExpectAndReturn`